### DISCUSSION TITLE
About Write()

### DISCUSSION THREAD

**Discussion by overflow218** [ANLAB-KAIST/KENSv3#81] [Project Q&A 2. Transport Layer (TCP)] [answered]
Dear TA, I'm trying to implement Write() systemcall. 
I sended data packet to peer in write(), but the peer does not respond for my data packet. 
peer doesn't also read my data packet. so I compared my code's pcap file and solution code's pcap file. 
this two are exactly same except for solution code get a respond from peer and mine is not. 
Is there any special reason that my peer does not read my data packet? (so I don't get a ACK packet from peer?) 
the test case looks like have a order like this
write(mine) -> close(mine) -> read(peer). 
Is it because mine is close before read(peer) so that mine does not get a packet from peer? 
Thank you!

---

안녕하세요 조교님 
write() 시스템콜을 구현하기 위해 시도중입니다. 
write()콜 안에서 상대방에게 데이터 패킷을 보냈는데, 상대방으로부터 ACK 패킷 응답이 오지 않고 상대방이 읽지도 못하는 상황입니다. 
pcap 파일을 통해 패킷을 살펴보았을땐 패킷 내용이 잘 채워져서 보내지는걸 확인했습니다. 혹시 상대방으로부터 응답을 받기 위한 
특별한 조건이 있는지 궁금합니다.  상대방이 read() 콜을 하기전에 제쪽 프로세스가 먼저 close() 하도록 테스트 케이스가 짜여져 있는것 같은데 
혹시 이때문인가요?
감사합니다.

**Comment by Pusnow** [comment, marked as answer]
> write()콜 안에서 상대방에게 데이터 패킷을 보냈는데, 상대방으로부터 ACK 패킷 응답이 오지 않고 상대방이 읽지도 못하는 상황입니다.

1. check TCP checksum using Wireshark. https://wiki.wireshark.org/TCP_Checksum_Verification
2. turn on (uncomment) logging features to see what happens in the solution binary.

https://github.com/ANLAB-KAIST/KENSv3/blob/22214b8c65fd01ec387634679626448702e6cb06/src/Networking/E_NetworkLog.cpp#L13-L25

> Is it because mine is close before read(peer) so that mine does not get a packet from peer?

No, at `FIN-WAIT-1` and `FIN-WAIT-2` states, the peer can read data.

**Reply by overflow218** [reply]
Thank you!
